solidity/contracts/MezoAllocator.sol

```diff
@@ -225,20 +231,32 @@ contract MezoAllocator is IDispatcher, Ownable2StepUpgradeable {
     function withdraw(uint256 amount) external {
         if (msg.sender != address(stbtc)) revert CallerNotStbtc();
 
-        emit DepositWithdrawn(depositId, amount);
+        uint256 unallocatedBalance = tbtc.balanceOf(address(this));
+
+        if (amount > unallocatedBalance) {
+            uint256 amountToWithdraw = amount - unallocatedBalance;
+
+            emit DepositWithdrawn(depositId, amountToWithdraw);
+
+            if (amountToWithdraw < depositBalance) {
+                mezoPortal.withdrawPartially(
+                    address(tbtc),
+                    depositId,
+                    uint96(amountToWithdraw)
+                );
+            } else if (amountToWithdraw > depositBalance) {
+                revert WithdrawalAmountExceedsDepositBalance(
+                    amountToWithdraw,
+                    depositBalance
+                );
+            } else {
+                mezoPortal.withdraw(address(tbtc), depositId);
+            }
 
-        if (amount < depositBalance) {
-            mezoPortal.withdrawPartially(
-                address(tbtc),
-                depositId,
-                uint96(amount)
-            );
-        } else {
-            mezoPortal.withdraw(address(tbtc), depositId);
+            // slither-disable-next-line reentrancy-no-eth
+            depositBalance -= uint96(amountToWithdraw);
         }
 
-        // slither-disable-next-line reentrancy-no-eth
-        depositBalance -= uint96(amount);
         tbtc.safeTransfer(address(stbtc), amount);
```

Comment:
Shouldn't we emit some event informing about withdrawing from the allocator? `DepositWithdrawn` is only about withdrawing from the Portal.

Reply:
dcbe752a3593b3d4af352a61e76926bb0dd0d3e6